text/0000-variadic-tuples.md

```diff
@@ -0,0 +1,1006 @@
+
+Declaration examples:
+
+- `struct VariadicStruct<(..T1)>` : declares a struct with a variadic tuple type identified by `T1` in its generic parameters
+- `fn my_func<(..T1), (..T2)>()`: a function can have variadic tuple type parameters
+- `impl<(..Head)>`: is an implementation block that uses a variadic tuple type identified by `Head`
+- `impl<A, B, C, (.._Tail)>`:  same as above, but with other generic parameters
+- `impl<A, B, (..C), (..D)>`: there can be several variadic tuple types declared in a generic parameter group
+- `impl<A, B, (..(C, D)), (..E)>`: and we can mix both syntaxes
+
+Usage examples:
+
+```rust
+struct VariadicStruct<(..T)>
+VariadicStruct<(usize,)>                // => (..T) matches (usize,)
+VariadicStruct<(usize, bool)>   // => (..T) matches (usize, bool)
+
+impl <(..(T1, T2))> VariadicStruct<(..(T1, T2))> { ... }
+VariadicStruct::<((usize, bool),)>
+// (..(T1, T2)) matches ((usize, bool),)
+// (..T1) is (usize,)
+// (..T2) is (bool,)
+VariadicStruct::<((usize, bool), (String, i8))> // (..(T1, T2)) matches ((usize, bool), (String, i8))
+// (..T1) is (usize, String)
+// (..T2) is (bool, i8)
+```
+
+### Expansion
+
+The expansion syntax is: `..<expr(T1, T2, ..., Tn)>` where `<expr(T1, T2, ..., Tn)>` is an expression using the variadic tuple type identifiers `T1`, `T2`, ..., `Tn`.
+
+Note: The expression in an expansion form can be enclosed by parenthesis for clarity. Ex: `..(T: Clone)`.
+
+The expansion form is allowed in all places where a type is allowed and in `where` bounds.
+
+Examples:
+
+```rust
+type TuplesOfRef<'a, (..T)> = (..&'a T);
+TuplesOfRef<'b, (usize, bool)>; // = (&'b usize, &'b bool)
+
+struct MegaMap<(..(K, V))> {
+  maps: (..HashMap<K, V>),
+}
+// 
+// struct MegaMap<((usize, bool), (String, i8))> {
+//   maps: (HashMap<usize, bool>, HashMap<String, i8>),
+// }
+
+fn append<(..L), (..R)>(l: (..L), r: (..R)) -> (..L, ..R)
+where 
+    ..(L: 'static + Clone), 
+    ..(R: 'static + Clone), {
+    ...
+}
+
+//
+// fn append<(usize, Vec<bool>), (&'static str, u8, i16)>(
+//       l: (usize, Vec<bool>), 
+//       r: (&'static str, u8, i16)
+//   ) -> (usize, Vec<bool>, &'static str, u8, i16) {
+//   ...
+// }
+```
+
+Note: If an expansion syntax does not contains any variadic tuple type identifier, it resolves to the unit type `( )`.
+
+Note2: If an expansion syntax contains multiple variadic tuple type identifiers, they must all have been declared together with the syntax `( ..(T1, T2, ..., Tn))` to ensure they have the same arity.
+
+## Variadic tuple
+
+A _variadic tuple_ is a variable of a variadic tuple type.
+
+### Destructuring a variadic tuple
+
+A variadic tuple can be destructured to manipulate its members.
+
+There are 3 syntaxes possible to destructure a variadic tuple for a variadic tuple `(..T)`:
+
+1. `(v @ ..)` of variadic tuple type `(..T)`
+2. `(ref v @ ..)` of variadic tuple type `(..&T)`
+3. `(ref mut v @ ..)` of variadic tuple type `(..&mut T)`
```

Comment:
This honestly isn't the right explanation for this, imho. I think it's mostly just appropriate to say that `..` in tuple patterns now can be bound to a variable to get a variadic tuple. You can also bind by reference as need be, or it'll automatically reference into `(..&T)` or `(..&mut T)` if the original expression was of type `&(..T)` or `&mut (..T)`.

Reply:
I agree, as this would also help explain why `(a, v @ .., b)` works.

Reply:
Indeed, that is way more easier to understand that way.

I updated the section accordingly